arangod/Graph/Providers/ClusterProvider.cpp

```diff
@@ -0,0 +1,334 @@
+      _query->warnings().registerWarning(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND, lE->getVertexIdentifier().toString());
+      _opts.getCache()->cacheVertex(std::move(lE->getVertexIdentifier()), VPackSlice::nullSlice());
+    }
+    result.emplace_back(std::move(lE));
+  }
+}
+
+Result ClusterProvider::fetchEdgesFromEngines(VertexType const& vertex) {
+  // Currently unused. This is just here as a placeholder. Needs to be implemented in the future
+  // as soon as we enable the ClusterProvider globally.
+  auto const* engines = _opts.getCache()->engines();
+
+  // TRI_ASSERT(vertexId.isString() || vertexId.isArray());
+
+  transaction::BuilderLeaser leased(trx());
+  leased->openObject(true);
+  leased->add("backward", VPackValue(_opts.isBackward()));
+  leased->add("keys", VPackValue(vertex.toString()));
+  leased->close();
+
+  auto* pool = trx()->vocbase().server().getFeature<NetworkFeature>().pool();
+
+  network::RequestOptions reqOpts;
+  reqOpts.database = trx()->vocbase().name();
+  reqOpts.skipScheduler = true;  // hack to avoid scheduler queue
+
+  std::vector<Future<network::Response>> futures;
+  futures.reserve(engines->size());
+
+  for (auto const& engine : *engines) {
+    futures.emplace_back(
+        network::sendRequestRetry(pool, "server:" + engine.first, fuerte::RestVerb::Put,
+                                  ::edgeUrl + StringUtils::itoa(engine.second),
+                                  leased->bufferRef(), reqOpts));
+  }
+
+  for (Future<network::Response>& f : futures) {
+    network::Response const& r = f.get();
+
+    if (r.fail()) {
+      return network::fuerteToArangoErrorCode(r);
+    }
+
+    auto payload = r.response().stealPayload();
+    VPackSlice resSlice(payload->data());
+    if (!resSlice.isObject()) {
+      // Response has invalid format
+      return TRI_ERROR_HTTP_CORRUPTED_JSON;
+    }
+    Result res = network::resultFromBody(resSlice, TRI_ERROR_NO_ERROR);
+    if (res.fail()) {
+      return res;
+    }
+    _stats.addFiltered(Helper::getNumericValue<size_t>(resSlice, "filtered", 0));
+    _stats.addScannedIndex(Helper::getNumericValue<size_t>(resSlice, "readIndex", 0));
+
+    bool allCached = true;
+    VPackSlice edges = resSlice.get("edges");
+    for (VPackSlice e : VPackArrayIterator(edges)) {
+      VPackSlice id = e.get(StaticStrings::IdString);
+      if (!id.isString()) {
+        // invalid id type
+        LOG_TOPIC("eb7cd", ERR, Logger::GRAPHS)
+            << "got invalid edge id type: " << id.typeName();
+        continue;
+      }
+
+      arangodb::velocypack::HashedStringRef edgeIdRef(id);
+
+      if (!_opts.getCache()->isEdgeCached(edgeIdRef, _opts.isBackward())) {
+        allCached = false;
+        _opts.getCache()->cacheEdge(vertex, edgeIdRef, e, _opts.isBackward());
+      }
+    }
+
+    if (!allCached) {
+      _opts.getCache()->datalake().add(std::move(payload));
+    }
+  }
+
+  return {};
```

Comment:
```suggestion
  return TRI_ERROR_NO_ERROR;
```

I would prefer this, as it is easier to understand that no error has happened, otherwise one needs to know the Default type,